### PR TITLE
Feature/929 stansummary cli11

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submisison Checklist

- [x] Run tests: `./runCmdStanTests.py src/test`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

Replaced use of `boost::program_options` with lib [CLI11](https://github.com/CLIUtils/CLI11),
which is a header-only C++11 library with appropriate open source license.

#### Intended Effect:

Pay down technical debt:

- Simplify build process.   Library `boost::program_options` is a non-header only library,
it made the build process longer, greatly increased the amount of the console output,

- Simplify makefiles

#### How to Verify:
Run unit tests

#### Side Effects:
NA

#### Documentation:
NA
#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):  Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: src/test/interface/stansummary_test.cpp [hunk 1]
```diff
@@ -411,7 +413,7 @@ TEST(CommandStansummary, check_csv_output) {
                                 + "interface" + path_separator
                                 + "example_output" + path_separator
                                 + "tmp_test_target_csv_file.csv";
-  std::string arg_csv_file = "--csv_file=" + target_csv_file;
```
Comment (bbbales2) [inline src/test/interface/stansummary_test.cpp:414]: Where does the `csv_file` -> `csv_filename` change come from?

**Review by bbbales2 — COMMENTED** [review]
The only changes here are the library, right? So it's the same features as the existing stansummary, just different code?

File: src/test/interface/stansummary_test.cpp (hunk 1 above)
Comment (mitzimorris) [reply src/test/interface/stansummary_test.cpp:414]: it was a bug in the test - the option long name is `csv_filename` - maybe the boost library was ok with prefix?

**Comment by mitzimorris** [comment]
> The only changes here are the library, right? So it's the same features as the existing stansummary, just different code?

correct - the CLI11 library is a better version of boost::program_options.   updated the code to use the library; changed the unit tests to match messages generated by CLI11.

File: src/cmdstan/stansummary.cpp [hunk 2]
```diff
@@ -36,131 +36,68 @@ Example:  stansummary model_chain_1.csv model_chain_2.csv
...
+  CLI::App app{"Allowed options"};
+  app.add_option("--sig_figs,-s", sig_figs, "Significant figures, default 2.",
+                 true)
+      ->check(CLI::Range(1, 18));
+  app.add_option("--autocorr,-a", autocorr_idx,
+                 "Display the chain autocorrelation.", true)
+      ->check(CLI::PositiveNumber);
+  app.add_option("--csv_filename,-c", csv_filename,
+                 "Write statistics to a csv.", true)
+      ->check(CLI::NonexistentPath);
```
Comment (bbbales2) [inline src/cmdstan/stansummary.cpp:60]: This checks that the output file is in a writeable directory?

File: src/cmdstan/stansummary.cpp [hunk 3]
```diff
@@ -36,131 +36,68 @@ Example:  stansummary model_chain_1.csv model_chain_2.csv
...
+  app.add_option("--percentiles,-p", percentiles_spec, "Percentiles to report.",
+                 true);
+  app.add_option("input_files", filenames, "Sampler csv files.", true)
+      ->required()
+      ->check(CLI::ExistingFile);
```
Comment (bbbales2) [inline src/cmdstan/stansummary.cpp:65]: This checks the input files are readable?

File: src/test/utility.hpp [hunk 4]
```diff
@@ -131,7 +131,8 @@ run_command_output run_command(std::string command) {
   using boost::posix_time::microsec_clock;
   using boost::posix_time::ptime;
   FILE *in;
-  in = popen(command.c_str(), "r");
+  std::string command_plus = command + " 2>&1";  // put stderr to stdout
```
Comment (bbbales2) [inline src/test/utility.hpp:134]: Why does this pipe stderr to stdout?

File: src/cmdstan/stansummary_helper.hpp [hunk 5]
```diff
@@ -301,11 +300,9 @@ Eigen::VectorXd percentiles_to_probs(
...
       cur_pct = pct;
     } catch (const std::exception &e) {
-      std::stringstream message_stream("");
-      message_stream << "position " << i << ", value " << percentiles[i] << ". "
-                     << "Values must be in range (1,99), inclusive, "
-                     << "and strictly increasing.";
-      throw boost::program_options::error(message_stream.str());
+      throw std::invalid_argument(
+          "values must be in range (1,99)"
+          ", inclusive, and strictly increasing.");
```
Comment (bbbales2) [inline src/cmdstan/stansummary_helper.hpp:305]: I think there needs to be a newline after the error. Dunno where to add this to the code, but when I hit the error my cursor ends up way over at the right:

```
bbales2@tadpole:~/cmdstan-develop$ ./bin/stansummary --percentiles=-10,3,4 output2.csv
Option --percentiles -10,3,4: values must be in range (1,99), inclusive, and strictly increasing.bbales2@tadpole:~/cmdstan-develop$ 
```

File: src/test/interface/stansummary_test.cpp [hunk 6]
```diff
@@ -289,14 +289,15 @@ TEST(CommandStansummary, bad_sig_figs_arg) {
...
+  expected_message = "--sig_figs: Value 101 not in range";
   out = run_command(command + " " + arg_sig_figs + " " + csv_file);
   EXPECT_TRUE(boost::algorithm::contains(out.output, expected_message));
   ASSERT_TRUE(out.hasError)
       << "\"" << out.command << "\" failed to quit with an error";
 }
 
 TEST(CommandStansummary, bad_autocorr_arg) {
-  std::string expected_message = "Bad value for option --autocorr";
+  std::string expected_message = "--autocorr: Number less or equal to 0";
```
Comment (bbbales2) [inline src/test/interface/stansummary_test.cpp:300]: If I don't enter an integer I get a not-super informative error:

```
bbales2@tadpole:~/cmdstan-develop$ ./bin/stansummary --autocorr=0.1 output2.csv
Could not convert: --autocorr = 0.1
Run with --help for more information.
```

Is there a way to get a better one? If not it's fine.

File: src/cmdstan/stansummary_helper.hpp [hunk 7]
```diff
@@ -332,6 +329,12 @@ stan::mcmc::chains<> parse_csv_files(const std::vector<std::string> &filenames,
   std::ifstream ifstream;
   ifstream.open(filenames[0].c_str());
   stan::io::stan_csv stan_csv = stan::io::stan_csv_reader::parse(ifstream, out);
+  if (stan_csv.samples.rows() < 1) {
```
Comment (bbbales2) [inline src/cmdstan/stansummary_helper.hpp:332]: I deleted all the samples in a file and got a segfault:

```
bbales2@tadpole:~/cmdstan-develop$ ./bin/stansummary output2.csv
Warning: non-fatal error reading samples
terminate called after throwing an instance of 'std::invalid_argument'
  what():  No sampling draws found in Stan CSV file: output2.csv.
Aborted (core dumped)
```

Comment (bbbales2) [reply src/cmdstan/stansummary_helper.hpp:332]: Similarly if I deleted the csv header I get a segfault:

```
bbales2@tadpole:~/cmdstan-develop$ ./bin/stansummary output3.csv
Warning: non-fatal error reading metadata
Error: error reading header
terminate called after throwing an instance of 'std::invalid_argument'
  what():  Error with header of input file in parse
Aborted (core dumped)
```

Seems like the errors are getting detected but there's still something weird happening with memory.

**Review by bbbales2 — CHANGES_REQUESTED** [review]
Review!

File: src/cmdstan/stansummary.cpp (hunk 2 above)
Comment (mitzimorris) [reply src/cmdstan/stansummary.cpp:60]: line 68 just checks that you're not clobbering an existing file. 
lines 92-100 check that the file is writeable.

looks like C++17 will have package "filesystem" that will do a nicer job on these checks.

File: src/cmdstan/stansummary.cpp (hunk 3 above)
Comment (mitzimorris) [reply src/cmdstan/stansummary.cpp:65]: checks that they exist.
need to add checks for readibility, similar to check for writable output csv file.

File: src/test/utility.hpp (hunk 4 above)
Comment (mitzimorris) [reply src/test/utility.hpp:134]: see https://stackoverflow.com/questions/6900577/c-popen-wont-catch-stderr
CLI11 lib writes arg parse errors to stderr; in order to check error message, we need stderr output as well.

File: src/cmdstan/stansummary.cpp (hunk 3 above)
Comment (mitzimorris) [reply src/cmdstan/stansummary.cpp:65]: done

File: src/cmdstan/stansummary_helper.hpp (hunk 5 above)
Comment (mitzimorris) [reply src/cmdstan/stansummary_helper.hpp:305]: leaving this message the same, added newline at point where error is reported - L 89 stansummary.cpp

File: src/test/interface/stansummary_test.cpp (hunk 6 above)
Comment (mitzimorris) [reply src/test/interface/stansummary_test.cpp:300]: excellent question.
looks like CLI11 "master" branch has additional "TypeValidator" but not yet released.  fuck'em.
could implement a workaround - assign to string, check positive int.  PITA.

File: src/cmdstan/stansummary_helper.hpp (hunk 7 above)
Comment (mitzimorris) [reply src/cmdstan/stansummary_helper.hpp:332]: good catch.  looking into it now.

File: src/cmdstan/stansummary.cpp (hunk 3 above)
Comment (bbbales2) [reply src/cmdstan/stansummary.cpp:65]: Totally optional, but when I get the error for readability I don't get the friendly `--help` prompt:

```
bbales2@tadpole:~/cmdstan-develop$ bin/stansummary output4.csv
Cannot read input csv file: output4.csv.
```

Compare this to:

```
bbales2@tadpole:~/cmdstan-develop$ bin/stansummary output12321.csv
input_files: File does not exist: output12321.csv
Run with --help for more information.
```

Comment (bbbales2) [reply src/cmdstan/stansummary.cpp:65]: Also I don't get the --help prompt with the percentiles option:

```
bbales2@tadpole:~/cmdstan-develop$ bin/stansummary -p -5 output4.csv
Option --percentiles -5: values must be in range (1,99), inclusive, and strictly increasing.
```

Very optional to add this.

File: src/test/interface/stansummary_test.cpp (hunk 6 above)
Comment (bbbales2) [reply src/test/interface/stansummary_test.cpp:300]: Skip it is fine with me. `--help` for more information is good enuff.

File: src/cmdstan/stansummary_helper.hpp (hunk 7 above)
Comment (mitzimorris) [reply src/cmdstan/stansummary_helper.hpp:333]: OK, there was an uncaught exception.  tightened up file open/close, added more checks.

**Comment by mitzimorris** [comment]
hi Ben, thanks for the thorough arg checking - this code should be far more robust.

also extremely useful w/r/t this PR:  https://github.com/stan-dev/cmdstan/pull/925
there's a lot more checks that could/should be done for that PR, but it would be cool to get it in -
not sure I'll be able to get to it for this release.

**Comment by mitzimorris** [comment]
@bbbales2 - ready for re-review

File: src/cmdstan/stansummary.cpp (hunk 3 above)
Comment (mitzimorris) [reply src/cmdstan/stansummary.cpp:65]: can't figure out how to plug in custom help message via CLI11 - spent several hours looking.
will give it another go.

Comment (bbbales2) [reply src/cmdstan/stansummary.cpp:65]: Eh if it's not easy let's not worry about it.

**Review by bbbales2 — APPROVED** [review]
Looks good to me. Merge if you're happy with it. I don't think it's worth fussing over the `--help` thing if you've already spent time trying to get that working.

File: src/cmdstan/stansummary.cpp (hunk 3 above)
Comment (mitzimorris) [reply src/cmdstan/stansummary.cpp:65]: doable but not easy.  worth doing as part of next PR.

**Pull request merged**